qualtran/bloqs/block_encoding.py

```diff
@@ -0,0 +1,272 @@
+        prepare: The bloq following the `Prepare` interface to wrap.
+        adjoint: Whether this is the adjoint preparation.
+    """
+
+    prepare: PrepareOracle
+    adjoint: bool = False
+
+    @cached_property
+    def selection_bitsize(self):
+        return sum(reg.total_bits() for reg in self.prepare.selection_registers)
+
+    @cached_property
+    def junk_bitsize(self):
+        return sum(reg.total_bits() for reg in self.prepare.selection_registers)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [Register('selection', self.selection_bitsize), Register('junk', self.junk_bitsize)]
+        )
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', selection: 'SoquetT', junk: 'SoquetT'
+    ) -> Dict[str, 'SoquetT']:
+        sel_regs = self.prepare.selection_registers
+        sel_part = Partition(self.selection_bitsize, regs=sel_regs)
+        sel_out_regs = bb.add_t(sel_part, x=selection)
+        jnk_regs = tuple(self.prepare.junk_registers)
+        jnk_part = Partition(self.junk_bitsize, regs=jnk_regs)
+        jnk_out_regs = bb.add_t(jnk_part, x=junk)
+        out_regs = bb.add(
+            self.prepare,
+            **{reg.name: sp for reg, sp in zip(sel_regs, sel_out_regs)},
+            **{reg.name: sp for reg, sp in zip(jnk_regs, jnk_out_regs)},
+        )
+        sel_out_regs = out_regs[: len(sel_regs)]
+        jnk_out_regs = out_regs[len(sel_regs) :]
+        selection = bb.add(
+            sel_part.dagger(), **{reg.name: sp for reg, sp in zip(sel_regs, sel_out_regs)}
+        )
+        junk = bb.add(
+            jnk_part.dagger(), **{reg.name: sp for reg, sp in zip(jnk_regs, jnk_out_regs)}
+        )
+        return {'selection': selection, 'junk': junk}
+
+    def dagger(self) -> 'BlackBoxPrepare':
+        return attrs.evolve(self, adjoint=not self.adjoint)
+
+    def short_name(self) -> str:
+        dag = '†' if self.adjoint else ''
+        return f'Prep{dag}'
+
+
+@attrs.frozen
+class BlackBoxBlockEncoding(BlockEncoding):
+    r"""Standard block encoding using SELECT and PREPARE.
+
+    Builds the block encoding via
+    $$
+        \mathcal{B}[V] = \mathrm{PREPARE}^\dagger \cdot \mathrm{SELECT} \cdot \mathrm{PREPARE},
+    $$
+    where
+    $$
+        \mathrm{PREPARE} |0\rangle_a = \sum_l \sqrt{\frac{w_l}{\lambda}} |l\rangle_a,
+    $$
+    and
+    $$
+        \mathrm{SELECT} |l\rangle_a|\psi\rangle_s = |l\rangle_a U_l |\psi\rangle_s.
+    $$
+    """
+
+    select: BlackBoxSelect
+    prepare: BlackBoxPrepare
+    adjoint: bool = False
+
+    @property
+    def selection_register(self) -> Register:
+        return Register('selection', bitsize=self.prepare.selection_bitsize)
+
+    @property
+    def junk_register(self) -> Register:
+        return Register('junk', bitsize=self.prepare.junk_bitsize)
+
+    @property
+    def system_register(self) -> Register:
+        return Register('system', bitsize=self.select.system_bitsize)
```

Comment:
Can you explain and motivate the design decision for having the abstract base construction for `BlockEncoding`. You could imagine un-abstracting these properties and make them always be defined like this. Instead your "api" would be that select and prepare need to expose a selection, junk, and system bitsize.

Reply:
My idea here was that if we want more complicated block encodings (i.e. add, chebychev, multiply etc which are given in Eq. 12-15 and Eq 33 above) these are also block encodings but should define the same interface as the "arguments" right?

I was also considering the case were say we have some complicated block encoding of the hamiltonian and maybe we don't want to worry about all the details but we want B[T_10[H]], then something like

``` python
@attrs.frozen
class ChebyshevPolynomial(BlockEncoding):
    r"""Block encoding of $T_j[H]$ where $T_j$ is the $j$-th Chebyshev polynomial.

    Here H is a Hamiltonian with spectral norm $|H| \le 1$ and we assume we have
    an $n_L$ qubit ancilla register and assume that $j > 0$ to avoid block
    encoding the identity operator.

    Recall:

    \begin{align*}
        T_0[H] &= \mathbb{1} \\
        T_1[H] &= H \\
        T_2[H] &= 2 H^2 - \mathbb{1} \\
        T_3[H] &= 4 H^3 - 3 H \\
        &\dots
    \end{align*}

    Args:
        block_encoding: Block encoding of a Hamiltonian $H$, $\mathcal{B}[H]$.
        order: order of Chebychev polynomial.

    References:
        [Quantum computing enhanced computational catalysis](
            https://arxiv.org/abs/2007.14460). Page 45; Theorem 1.
    """

    block_encoding: BlockEncoding
    order: int

    def __attrs_post_init__(self):
        if self.order < 1:
            raise ValueError(f"order must be greater >= 1. Found {self.order}.")

    def short_name(self) -> str:
        return f"T_{self.order}[{self.block_encoding.short_name()}]"

    @property
    def selection_register(self) -> Register:
        return Register('selection', bitsize=self.block_encoding.selection_bitsize)

    @property
    def junk_register(self) -> Register:
        return Register('junk', bitsize=self.block_encoding.junk_bitsize)

    @property
    def system_register(self) -> Register:
        return Register('system', bitsize=self.block_encoding.system_bitsize)

    def build_composite_bloq(
        self, bb: 'BloqBuilder', selection: 'SoquetT', junk: 'SoquetT', system: 'SoquetT'
    ) -> Dict[str, 'Soquet']:
        # includes selection registers and any selection registers used by PREPARE
        selection, junk, system = bb.add(
            self.block_encoding, selection=selection, junk=junk, system=system
        )
        for iorder in range(1, self.order):
            selection = bb.add(Reflection(bitsize=self.selection_bitsize), refl=selection)
            be = attrs.evolve(self.block_encoding, adjoint=((iorder % 2) == 1))
            selection, junk, system = bb.add(be, selection=selection, junk=junk, system=system)
        return {'selection': selection, 'junk': junk, 'system': system}

    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
        n = self.order
        num_refl = self.block_encoding.selection_bitsize
        return {(Reflection(bitsize=num_refl), n - 1), (self.block_encoding, n)}


```

Reply:
but in your answer, you still have the same register names; albeit with a different source of the bitsizes